File: server/typescript/packages/metadata/src/core/object/validate-discriminator.ts
// Validation pass: TPH discriminator cross-attribute rules (FR-014).
//
// Codes (all errors):
//   ERR_DISCRIMINATOR_FIELD_NOT_FOUND      — @discriminator names a field that
//                                            does not exist on the entity (own
//                                            or via extends chain).
//   ERR_DISCRIMINATOR_VALUE_DUPLICATE      — two subtypes of the same
//                                            @discriminator-bearing root claim
//                                            the same @discriminatorValue.
//   ERR_DISCRIMINATOR_VALUE_MISSING        — a concrete (non-abstract) entity
//                                            extends a chain whose root carries
//                                            @discriminator but the subtype lacks
//                                            @discriminatorValue.
//   ERR_DISCRIMINATOR_VALUE_TYPE_MISMATCH  — @discriminatorValue cannot be
//                                            coerced to the discriminator field's
//                                            subtype (enum: not in @values;
//                                            int: not numeric; string: always OK).

import type { MetaData } from "../../shared/meta-data.js";
import { ParseError } from "../../errors.js";
import { TYPE_OBJECT, TYPE_FIELD } from "../../shared/base-types.js";
import {
  OBJECT_ATTR_DISCRIMINATOR,
  OBJECT_ATTR_DISCRIMINATOR_VALUE,
  OBJECT_SUBTYPE_ENTITY,
} from "./object-constants.js";
import {
  FIELD_SUBTYPE_ENUM,
  FIELD_SUBTYPE_INT,
  FIELD_SUBTYPE_LONG,
  FIELD_SUBTYPE_STRING,
  FIELD_ATTR_VALUES,
} from "../field/field-constants.js";

const NUMERIC_DISCRIMINATOR_SUBTYPES = new Set<string>([
  FIELD_SUBTYPE_INT,
  FIELD_SUBTYPE_LONG,
]);

export function validateDiscriminator(root: MetaData): ParseError[] {
  const errors: ParseError[] = [];

  const entities = root
    .ownChildren()
    .filter((c) => c.type === TYPE_OBJECT && c.subType === OBJECT_SUBTYPE_ENTITY);

  // Pass 1: @discriminator on bases — name resolution (own + inherited fields).
  for (const obj of entities) {
    const disc = obj.ownAttr(OBJECT_ATTR_DISCRIMINATOR);
    if (typeof disc !== "string" || disc === "") continue;

    const field = findFieldOnEntity(obj, disc);
    if (field === undefined) {
      errors.push(
        new ParseError(
          `object.entity "${obj.name}" @discriminator: "${disc}" does not name a field on this entity ` +
            `(checked own children and the extends chain)`,
          { code: "ERR_DISCRIMINATOR_FIELD_NOT_FOUND", source: obj.source },
        ),
      );
    }
  }

  // Pass 2: @discriminatorValue validation — for every entity that declares it,
  // walk up to find the @discriminator-bearing root, then check field membership
  // and uniqueness within that root's subtype set.
  type SubtypeBinding = {
    subtype: MetaData;
    value: string;
    discField: MetaData;
  };
  const bindingsByRoot = new Map<MetaData, SubtypeBinding[]>();

  for (const obj of entities) {
    const value = obj.ownAttr(OBJECT_ATTR_DISCRIMINATOR_VALUE);
    if (typeof value !== "string" || value === "") continue;

    const { root: discRoot, fieldName } = findDiscriminatorRoot(obj);
    if (discRoot === undefined || fieldName === undefined) continue;

    const field = findFieldOnEntity(discRoot, fieldName);
    if (field === undefined) continue; // root's own ERR_DISCRIMINATOR_FIELD_NOT_FOUND already fires

    // Type-match check.
    if (field.subType === FIELD_SUBTYPE_ENUM) {
      const enumValues = field.ownAttr(FIELD_ATTR_VALUES);
      const list = Array.isArray(enumValues) ? enumValues.map(String) : [];
      if (!list.includes(value)) {
        errors.push(
          new ParseError(
            `object.entity "${obj.name}" @discriminatorValue: "${value}" is not a member of the ` +
              `discriminator enum field "${fieldName}" @values [${list.join(", ")}]`,
            { code: "ERR_DISCRIMINATOR_VALUE_TYPE_MISMATCH", source: obj.source },
          ),
        );
      }
    } else if (NUMERIC_DISCRIMINATOR_SUBTYPES.has(field.subType)) {
      if (!/^-?\d+$/.test(value)) {
        errors.push(
          new ParseError(
            `object.entity "${obj.name}" @discriminatorValue: "${value}" does not coerce to ` +
              `numeric discriminator field "${fieldName}" (field.${field.subType})`,
            { code: "ERR_DISCRIMINATOR_VALUE_TYPE_MISMATCH", source: obj.source },
          ),
        );
      }
    } else if (field.subType !== FIELD_SUBTYPE_STRING) {
      // Discriminator fields outside {enum, integer-family, string} are
      // unusual; accept silently (a future refinement may add a warning).
    }

    let list = bindingsByRoot.get(discRoot);
    if (list === undefined) {
      list = [];
      bindingsByRoot.set(discRoot, list);
    }
    list.push({ subtype: obj, value, discField: field });
  }

  // Pass 3: ERR_DISCRIMINATOR_VALUE_DUPLICATE within each root's subtypes.
  for (const [, bindings] of bindingsByRoot) {
    const seen = new Map<string, MetaData>();
    for (const b of bindings) {
      const prev = seen.get(b.value);
      if (prev !== undefined) {
        errors.push(
          new ParseError(
            `object.entity "${b.subtype.name}" @discriminatorValue: "${b.value}" ` +
              `duplicates the value already claimed by "${prev.name}"`,
            { code: "ERR_DISCRIMINATOR_VALUE_DUPLICATE", source: b.subtype.source },
          ),
        );
      } else {
        seen.set(b.value, b.subtype);
      }
    }
  }

  // Pass 4: ERR_DISCRIMINATOR_VALUE_MISSING — every concrete (non-abstract)
  // entity that extends a @discriminator-bearing root must declare a value.
  for (const obj of entities) {
    if (obj.isAbstract === true) continue;
    if (typeof obj.ownAttr(OBJECT_ATTR_DISCRIMINATOR_VALUE) === "string") continue;
    if (typeof obj.ownAttr(OBJECT_ATTR_DISCRIMINATOR) === "string") continue; // a root, not a subtype
    const { root: discRoot } = findDiscriminatorRoot(obj);
    if (discRoot === undefined) continue;
    if (discRoot === obj) continue; // safety: same node, shouldn't happen
    errors.push(
      new ParseError(
        `object.entity "${obj.name}" extends the @discriminator-bearing root "${discRoot.name}" ` +
          `but is missing @discriminatorValue (required on every concrete subtype)`,
        { code: "ERR_DISCRIMINATOR_VALUE_MISSING", source: obj.source },
      ),
    );
  }

  return errors;
}

/** Find a field with the given name on `entity` — own first, then via the
 *  resolved extends chain. Returns the declaring field node (own attrs intact). */
function findFieldOnEntity(entity: MetaData, name: string): MetaData | undefined {
  for (const child of entity.ownChildren()) {
    if (child.type === TYPE_FIELD && child.name === name) return child;
  }
  let cursor = entity.superResolved;
  while (cursor !== undefined) {
    for (const child of cursor.ownChildren()) {
      if (child.type === TYPE_FIELD && child.name === name) return child;
    }
    cursor = cursor.superResolved;
  }
  return undefined;
}

/** Walk the extends chain to find the first ancestor (or self) carrying
 *  @discriminator. Returns {root, fieldName} when found; both undefined
 *  otherwise. */
function findDiscriminatorRoot(
  entity: MetaData,
): { root: MetaData | undefined; fieldName: string | undefined } {
  let cursor: MetaData | undefined = entity;
  while (cursor !== undefined) {
    const v = cursor.ownAttr(OBJECT_ATTR_DISCRIMINATOR);
    if (typeof v === "string" && v !== "") {
      return { root: cursor, fieldName: v };
    }
    cursor = cursor.superResolved;
  }
  return { root: undefined, fieldName: undefined };
}

File: server/python/src/metaobjects/loader/validate_discriminator.py
"""FR-014 — TPH discriminator cross-attribute rules.

Codes (all errors):
    * ``ERR_DISCRIMINATOR_FIELD_NOT_FOUND``     — ``@discriminator`` names a field
      that does not exist on the entity (own or via extends chain).
    * ``ERR_DISCRIMINATOR_VALUE_DUPLICATE``     — two subtypes of the same
      ``@discriminator``-bearing root claim the same ``@discriminatorValue``.
    * ``ERR_DISCRIMINATOR_VALUE_MISSING``       — a concrete (non-abstract) entity
      extends a chain whose root carries ``@discriminator`` but lacks
      ``@discriminatorValue``.
    * ``ERR_DISCRIMINATOR_VALUE_TYPE_MISMATCH`` — ``@discriminatorValue`` cannot be
      coerced to the discriminator field's subtype (enum: not in ``@values``;
      integer-family: not numeric; string: always OK).

Mirrors the TS reference
``packages/metadata/src/core/object/validate-discriminator.ts``.
"""
from __future__ import annotations

import re

from ..errors import ErrorCode, MetaError
from ..meta.meta_data import MetaData
from ..meta.core.field.field_constants import (
    FIELD_ATTR_VALUES,
    FIELD_SUBTYPE_ENUM,
    FIELD_SUBTYPE_INT,
    FIELD_SUBTYPE_LONG,
    FIELD_SUBTYPE_STRING,
)
from ..meta.core.object.object_constants import (
    OBJECT_ATTR_DISCRIMINATOR,
    OBJECT_ATTR_DISCRIMINATOR_VALUE,
    OBJECT_SUBTYPE_ENTITY,
)
from ..shared.base_types import TYPE_FIELD, TYPE_OBJECT

_NUMERIC_DISCRIMINATOR_SUBTYPES = frozenset(
    {FIELD_SUBTYPE_INT, FIELD_SUBTYPE_LONG}
)
_INT_RE = re.compile(r"^-?\d+$")


def _find_field_on_entity(entity: MetaData, name: str) -> MetaData | None:
    """A field with ``name`` on ``entity`` — own first, then via extends chain."""
    for child in entity.own_children():
        if child.type == TYPE_FIELD and child.name == name:
            return child
    cursor = entity.super_data
    while cursor is not None:
        for child in cursor.own_children():
            if child.type == TYPE_FIELD and child.name == name:
                return child
        cursor = cursor.super_data
    return None


def _find_discriminator_root(entity: MetaData) -> tuple[MetaData | None, str | None]:
    """First ancestor (or self) carrying ``@discriminator``: (root, fieldName)."""
    cursor: MetaData | None = entity
    while cursor is not None:
        v = cursor.attr(OBJECT_ATTR_DISCRIMINATOR)
        if isinstance(v, str) and v != "":
            return cursor, v
        cursor = cursor.super_data
    return None, None


def validate_discriminator(root: MetaData, errors: list[MetaError]) -> None:
    entities = [
        c
        for c in root.own_children()
        if c.type == TYPE_OBJECT and c.sub_type == OBJECT_SUBTYPE_ENTITY
    ]

    # Pass 1: @discriminator name resolution (own + inherited fields).
    for obj in entities:
        disc = obj.attr(OBJECT_ATTR_DISCRIMINATOR)
        if not isinstance(disc, str) or disc == "":
            continue
        if _find_field_on_entity(obj, disc) is None:
            errors.append(
                MetaError(
                    f'object.entity "{obj.name}" @discriminator: "{disc}" does not '
                    "name a field on this entity (checked own children and the "
                    "extends chain)",
                    ErrorCode.ERR_DISCRIMINATOR_FIELD_NOT_FOUND,
                    envelope=obj.source,
                )
            )

    # Pass 2: @discriminatorValue type-check + collect bindings per root.
    bindings_by_root: list[tuple[MetaData, list[tuple[MetaData, str]]]] = []
    _root_index: dict[int, list[tuple[MetaData, str]]] = {}

    for obj in entities:
        value = obj.attr(OBJECT_ATTR_DISCRIMINATOR_VALUE)
        if not isinstance(value, str) or value == "":
            continue

        disc_root, field_name = _find_discriminator_root(obj)
        if disc_root is None or field_name is None:
            continue
        field = _find_field_on_entity(disc_root, field_name)
        if field is None:
            continue  # root's own ERR_DISCRIMINATOR_FIELD_NOT_FOUND already fires

        if field.sub_type == FIELD_SUBTYPE_ENUM:
            enum_values = field.attr(FIELD_ATTR_VALUES)
            members = [str(v) for v in enum_values] if isinstance(enum_values, (list, tuple)) else []
            if value not in members:
                errors.append(
                    MetaError(
                        f'object.entity "{obj.name}" @discriminatorValue: "{value}" '
                        f'is not a member of the discriminator enum field '
                        f'"{field_name}" @values [{", ".join(members)}]',
                        ErrorCode.ERR_DISCRIMINATOR_VALUE_TYPE_MISMATCH,
                        envelope=obj.source,
                    )
                )
        elif field.sub_type in _NUMERIC_DISCRIMINATOR_SUBTYPES:
            if _INT_RE.match(value) is None:
                errors.append(
                    MetaError(
                        f'object.entity "{obj.name}" @discriminatorValue: "{value}" '
                        f'does not coerce to numeric discriminator field '
                        f'"{field_name}" (field.{field.sub_type})',
                        ErrorCode.ERR_DISCRIMINATOR_VALUE_TYPE_MISMATCH,
                        envelope=obj.source,
                    )
                )
        elif field.sub_type != FIELD_SUBTYPE_STRING:
            # Non-{enum, integer-family, string} discriminators accepted silently.
            pass

        existing = _root_index.get(id(disc_root))
        if existing is None:
            existing = []
            _root_index[id(disc_root)] = existing
            bindings_by_root.append((disc_root, existing))
        existing.append((obj, value))

    # Pass 3: ERR_DISCRIMINATOR_VALUE_DUPLICATE within each root's subtypes.
    for _disc_root, bindings in bindings_by_root:
        seen: dict[str, MetaData] = {}
        for subtype, value in bindings:
            prev = seen.get(value)
            if prev is not None:
                errors.append(
                    MetaError(
                        f'object.entity "{subtype.name}" @discriminatorValue: '
                        f'"{value}" duplicates the value already claimed by '
                        f'"{prev.name}"',
                        ErrorCode.ERR_DISCRIMINATOR_VALUE_DUPLICATE,
                        envelope=subtype.source,
                    )
                )
            else:
                seen[value] = subtype

    # Pass 4: ERR_DISCRIMINATOR_VALUE_MISSING — every concrete entity that extends
    # a @discriminator-bearing root must declare a value.
    for obj in entities:
        if obj.is_abstract is True:
            continue
        if isinstance(obj.attr(OBJECT_ATTR_DISCRIMINATOR_VALUE), str):
            continue
        if isinstance(obj.attr(OBJECT_ATTR_DISCRIMINATOR), str):
            continue  # a root, not a subtype
        disc_root, _ = _find_discriminator_root(obj)
        if disc_root is None or disc_root is obj:
            continue
        errors.append(
            MetaError(
                f'object.entity "{obj.name}" extends the @discriminator-bearing '
                f'root "{disc_root.name}" but is missing @discriminatorValue '
                "(required on every concrete subtype)",
                ErrorCode.ERR_DISCRIMINATOR_VALUE_MISSING,
                envelope=obj.source,
            )
        )
